<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="shiny.gosling">
<title>1. Introduction to shiny.gosling • shiny.gosling</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="1. Introduction to shiny.gosling">
<meta property="og:description" content="shiny.gosling">
<meta property="og:image" content="https://appsilon.github.io/shiny.gosling/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Google tag (gtag.js) --><script async src="https://www.googletagmanager.com/gtag/js?id=G-SESD3QWD7B"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-SESD3QWD7B');
</script><!-- Google tag (gtag.js) --><script async src="https://www.googletagmanager.com/gtag/js?id=G-FQQZL5V93G"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-FQQZL5V93G');
</script><script async defer src="https://buttons.github.io/buttons.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">shiny.gosling</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.2</small>

    <a href="https://shiny.tools" class="external-link home"><i class="fa fa-lg fa-chevron-left"></i></a>

    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home"></span>
     
    Start
  </a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/intro.html">
    <span class="fa fa-university"></span>
     
    Tutorial
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/GRanges.html">
    <span class="fa fa-university"></span>
     
    Using GRanges
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Reference
  </a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Appsilon/shiny.gosling">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/Appsilon">
    <span class="fa fa-twitter fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://fosstodon.org/@appsilon">
    <span class="fab fa fab fa-mastodon fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>1. Introduction to shiny.gosling</h1>
            
      
      
      <div class="d-none name"><code>intro.Rmd</code></div>
    </div>

    
    
<p>This is a Shiny app written in R that creates an interactive
visualization using the shiny and shiny.gosling packages. The
visualization displays genomic data from the Cistrome database in a
track-based layout.</p>
<div class="section level2">
<h2 id="section-call-library-libraries-">Call library libraries.<a class="anchor" aria-label="anchor" href="#section-call-library-libraries-"></a>
</h2>
<div class="sourceCode" id="section-cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">shiny.gosling</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://shiny.posit.co/" class="external-link">shiny</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r-lib/sessioninfo#readme" class="external-link">sessioninfo</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="section-introduction-to-shiny-gosling">Introduction to shiny.gosling<a class="anchor" aria-label="anchor" href="#section-introduction-to-shiny-gosling"></a>
</h2>
<p>This package is an R shiny implementation of the <a href="http://gosling-lang.org/" class="external-link"><strong>Gosling.js</strong></a> library.
Gosling js is a highly <em>expressive</em> library based on a
<em>Grammar for scalable and interactive Genomics Data
visualization</em>. This library is build upon the <a href="https://github.com/gosling-lang/gosling-react" class="external-link"><strong>React</strong>
wrapper of the Gosling.js</a>. Which is powered by <a href="https://github.com/gosling-lang/gosling-react" class="external-link"><strong>Shiny.React</strong></a>.
Thus most of the Charts can be directly build using this package.</p>
<p>Let’s start with an example. For that let’s read a csv file which has
some Genomic data with chromosome start stop indicators. This dataset is
directly picked from gosling-lang.org <a href="https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec" class="external-link">here</a>.</p>
<p><a href="https://docs.higlass.io/data_preparation.html#multivec-files" class="external-link">Multivec</a>
is a file format introduced by <a href="https://higlass.io/" class="external-link">Higlass</a>
suitable for representing and visualizing multi-dimensional numerical
data across genomic coordinates. It’s commonly used for representing
data like ChIP-seq, ATAC-seq, Hi-C, and other genomic experiments where
signals or measurements are collected at various genomic positions.</p>
<p>Multivec data is essentially a matrix where rows correspond to
different genomic positions or regions, and columns correspond to
different samples or experiments. Each entry in the matrix represents a
value associated with a specific genomic position and sample. The
genomic positions along the rows of the matrix are usually represented
as chromosomal coordinates (chromosome name and base pair position).
This allows the data to be aligned with the genome, enabling accurate
visualization and analysis. There are different tools and file formats
that support multivec data, allowing researchers to work with and
visualize this type of data. The bigWig and bedGraph formats are
commonly used for representing multivec data. Visualization tools and
libraries like the UCSC Genome Browser, IGV (Integrative Genomics
Viewer), and libraries like “shiny.gosling” can render multivec data
visualizations.</p>
<p>Here are some resources and links where you can learn more about
multivec data and how it’s used in genomics research:</p>
<p>UCSC Genome Browser:</p>
<p>The UCSC Genome Browser is a widely used tool for visualizing genomic
data, including multivec data. <a href="https://genome.ucsc.edu/training/" class="external-link">Tutorial on visualizing
multivec data in the UCSC Genome Browser</a></p>
<p>IGV (Integrative Genomics Viewer):</p>
<p>IGV is another popular genome visualization tool that supports
multivec data. <a href="https://www.youtube.com/watch?v=P9n0tZxiwPs" class="external-link">Tutorial on loading
and visualizing multivec data in IGV</a></p>
<p>BedGraph and BigWig Formats:</p>
<p>These are common file formats used for representing multivec data. <a href="https://genome.ucsc.edu/goldenPath/help/bedgraph.html" class="external-link">Explanation
of the BedGraph format</a> <a href="https://genome.ucsc.edu/goldenPath/help/bigWig.html" class="external-link">Explanation
of the BigWig format</a></p>
<div class="sourceCode" id="section-cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">cistrome_data</span> <span class="op">&lt;-</span> <span class="st">"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec"</span></span></code></pre></div>
<p>In shiny.gosling we can basically create tracks from data and then
create view from tracks. To understand how to build a plot let’s
understand 3 basic principles of <code>gosling.js</code></p>
<ol style="list-style-type: decimal">
<li>
<code>Track</code> contains data, layout, height, width and all
aesthetics etc…</li>
<li>One or more Track combine together create a <code>View</code>
</li>
<li>One or more views combine together create a <code>Plot</code>.</li>
</ol>
<p>This is how a plot is created in Gosling. Let’s visit this one by
one.</p>
<p>Let’s start by creating a track. Let’s define the first track and add
more properties to the track. With <code>shiny.gosling</code> you can be
specific about the colors and ranges and channel.</p>
<p>Let’s build the layers for the plot. So we can build multiple tracks
to represent the genome. <code>add_single_track</code> function
constructs a single track from the inputs.</p>
<p>The track_data function is used to define the data source for the
track. It specifies the URL of the dataset, the type of data
(“multivec”), and various data-related parameters such as rows, columns,
values, categories, and bin size.</p>
<p>The visualization of the track is specified using various visual
channels such as x, xe, row, color, and tooltip.</p>
<div class="sourceCode" id="section-cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_track</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_single_track.html">add_single_track</a></span><span class="op">(</span></span>
<span>  id <span class="op">=</span> <span class="st">"track1"</span>,</span>
<span>  data <span class="op">=</span> <span class="fu"><a href="../reference/track_data.html">track_data</a></span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="va">cistrome_data</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"multivec"</span>,</span>
<span>    row <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    column <span class="op">=</span> <span class="st">"position"</span>,</span>
<span>    value <span class="op">=</span> <span class="st">"peak"</span>,</span>
<span>    categories <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample 1"</span>, <span class="st">"sample 2"</span>, <span class="st">"sample 3"</span>, <span class="st">"sample 4"</span><span class="op">)</span>,</span>
<span>    binSize <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  <span class="op">)</span>,</span>
<span>  mark <span class="op">=</span> <span class="st">"rect"</span>,</span>
<span>  x <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_x.html">visual_channel_x</a></span><span class="op">(</span>field <span class="op">=</span> <span class="st">"start"</span>, type <span class="op">=</span> <span class="st">"genomic"</span>, axis <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span>,</span>
<span>  xe <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_x.html">visual_channel_x</a></span><span class="op">(</span>field <span class="op">=</span> <span class="st">"end"</span>, type <span class="op">=</span> <span class="st">"genomic"</span><span class="op">)</span>,</span>
<span>  row <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_row.html">visual_channel_row</a></span><span class="op">(</span></span>
<span>    field <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"nominal"</span>,</span>
<span>    legend <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span>,</span>
<span>  color <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_color.html">visual_channel_color</a></span><span class="op">(</span></span>
<span>    field <span class="op">=</span> <span class="st">"peak"</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"quantitative"</span>,</span>
<span>    legend <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span>,</span>
<span>  tooltip <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_tooltips.html">visual_channel_tooltips</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/visual_channel_tooltip.html">visual_channel_tooltip</a></span><span class="op">(</span>field <span class="op">=</span> <span class="st">"start"</span>, type <span class="op">=</span> <span class="st">"genomic"</span>, alt <span class="op">=</span> <span class="st">"Start Position"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="../reference/visual_channel_tooltip.html">visual_channel_tooltip</a></span><span class="op">(</span>field <span class="op">=</span> <span class="st">"end"</span>, type <span class="op">=</span> <span class="st">"genomic"</span>, alt <span class="op">=</span> <span class="st">"End Position"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="../reference/visual_channel_tooltip.html">visual_channel_tooltip</a></span><span class="op">(</span></span>
<span>      field <span class="op">=</span> <span class="st">"peak"</span>,</span>
<span>      type <span class="op">=</span> <span class="st">"quantitative"</span>,</span>
<span>      alt <span class="op">=</span> <span class="st">"Value"</span>,</span>
<span>      format <span class="op">=</span> <span class="st">"0.2"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  width <span class="op">=</span> <span class="fl">600</span>,</span>
<span>  height <span class="op">=</span> <span class="fl">130</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The compose_view function is used to compose the single track
visualization into a view. The composed view is placed within a circular
layout. The x-domain (genomic interval) is set to chromosome 1, interval
[1, 3000500].</p>
<div class="sourceCode" id="section-cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_composed_track</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compose_view.html">compose_view</a></span><span class="op">(</span></span>
<span>  tracks <span class="op">=</span> <span class="va">single_track</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="section-cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_composed_views</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/arrange_views.html">arrange_views</a></span><span class="op">(</span></span>
<span>  title <span class="op">=</span> <span class="st">"Single Track"</span>,</span>
<span>  subtitle <span class="op">=</span> <span class="st">"This is the simplest single track visualization with a linear layout"</span>,</span>
<span>  layout <span class="op">=</span> <span class="st">"circular"</span>,</span>
<span>  views <span class="op">=</span> <span class="va">single_composed_track</span>,</span>
<span>  xDomain <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    chromosome <span class="op">=</span> <span class="st">"chr1"</span>,</span>
<span>    interval <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3000500</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We can then even add more tracks to it. So let’s create a few more
tracks just to make a better and more beautiful graph.</p>
<p>The use_gosling function is used to incorporate the shiny.gosling
package for rendering the visualization.</p>
<p>Hovering on the plot will show the start position, the end position
and the value.</p>
<p>Scrolling on the plot, will zoom in or zoom out the view.</p>
<p>Users can click on the PDF button to download a pdf of the current
state of the plot.</p>
<div class="sourceCode" id="section-cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">server</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">input</span>, <span class="va">output</span>, <span class="va">session</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>  <span class="va">output</span><span class="op">$</span><span class="va">gosling_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderGosling.html">renderGosling</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/gosling.html">gosling</a></span><span class="op">(</span></span>
<span>      component_id <span class="op">=</span> <span class="st">"component_1"</span>,</span>
<span>      <span class="va">single_composed_views</span>,</span>
<span>      clean_braces <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">ui</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/shiny/man/fluidPage.html" class="external-link">fluidPage</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/use_gosling.html">use_gosling</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/goslingOutput.html">goslingOutput</a></span><span class="op">(</span><span class="st">"gosling_plot"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/shinyApp.html" class="external-link">shinyApp</a></span><span class="op">(</span><span class="va">ui</span>, <span class="va">server</span>, options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>height <span class="op">=</span> <span class="fl">700</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro.png" width="500"></p>
</div>
<div class="section level2">
<h2 id="section-session-info">Session Info<a class="anchor" aria-label="anchor" href="#section-session-info"></a>
</h2>
<div class="sourceCode" id="section-cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] vctrs_0.6.3       cli_3.6.1         knitr_1.44        rlang_1.1.1      </span></span>
<span><span class="co">#&gt;  [5] xfun_0.40         stringi_1.7.12    purrr_1.0.2       promises_1.2.1   </span></span>
<span><span class="co">#&gt;  [9] textshaping_0.3.6 shiny_1.7.5       xtable_1.8-4      jsonlite_1.8.7   </span></span>
<span><span class="co">#&gt; [13] glue_1.6.2        rprojroot_2.0.3   htmltools_0.5.6   httpuv_1.6.11    </span></span>
<span><span class="co">#&gt; [17] ragg_1.2.5        sass_0.4.7        rmarkdown_2.25    evaluate_0.21    </span></span>
<span><span class="co">#&gt; [21] jquerylib_0.1.4   ellipsis_0.3.2    fastmap_1.1.1     yaml_2.3.7       </span></span>
<span><span class="co">#&gt; [25] lifecycle_1.0.3   memoise_2.0.1     stringr_1.5.0     compiler_4.3.1   </span></span>
<span><span class="co">#&gt; [29] fs_1.6.3          Rcpp_1.0.11       later_1.3.1       systemfonts_1.0.4</span></span>
<span><span class="co">#&gt; [33] digest_0.6.33     R6_2.5.1          magrittr_2.0.3    bslib_0.5.1      </span></span>
<span><span class="co">#&gt; [37] tools_4.3.1       mime_0.12         pkgdown_2.0.7     cachem_1.0.8     </span></span>
<span><span class="co">#&gt; [41] desc_1.4.2</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed with <span class="emoji" data-emoji="heart">❤️</span> by <a href="https://appsilon.com" class="external-link">Appsilon</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
