<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="shiny.gosling">
<title>3. Creating a Circos Plot with VCF Data • shiny.gosling</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="3. Creating a Circos Plot with VCF Data">
<meta property="og:description" content="shiny.gosling">
<meta property="og:image" content="https://appsilon.github.io/shiny.gosling/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Google tag (gtag.js) --><script async src="https://www.googletagmanager.com/gtag/js?id=G-SESD3QWD7B"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-SESD3QWD7B');
</script><!-- Google tag (gtag.js) --><script async src="https://www.googletagmanager.com/gtag/js?id=G-FQQZL5V93G"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-FQQZL5V93G');
</script><script async defer src="https://buttons.github.io/buttons.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">shiny.gosling</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.2</small>

    <a href="https://shiny.tools" class="external-link home"><i class="fa fa-lg fa-chevron-left"></i></a>

    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home"></span>
     
    Start
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/intro.html">
    <span class="fa fa-university"></span>
     
    Tutorial
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/GRanges.html">
    <span class="fa fa-university"></span>
     
    Using GRanges
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    Reference
  </a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Appsilon/shiny.gosling">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/Appsilon">
    <span class="fa fa-twitter fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://fosstodon.org/@appsilon">
    <span class="fab fa fab fa-mastodon fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>3. Creating a Circos Plot with VCF Data</h1>
            
      
      
      <div class="d-none name"><code>VCF.Rmd</code></div>
    </div>

    
    
<p>This tutorial walks through the process of creating a Circos plot
using <a href="https://genome.ucsc.edu/goldenPath/help/vcf.html" class="external-link">VCF</a>
(Variant Call Format) data, with a focus on utilizing the shiny.gosling
package to achieve interactive visualizations within a Shiny app.</p>
<div class="section level2">
<h2 id="section-libraries-and-initialization">Libraries and Initialization<a class="anchor" aria-label="anchor" href="#section-libraries-and-initialization"></a>
</h2>
<p>First, we load the necessary libraries for our analysis, including
StructuralVariantAnnotation, ggbio, shiny, and shiny.gosling. We also
initialize the shiny.gosling package using the use_gosling()
function.</p>
<div class="sourceCode" id="section-cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va">StructuralVariantAnnotation</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://lawremi.github.io/ggbio/" class="external-link">ggbio</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://shiny.posit.co/" class="external-link">shiny</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va">shiny.gosling</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/use_gosling.html">use_gosling</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="section-loading-vcf-data">Loading VCF Data<a class="anchor" aria-label="anchor" href="#section-loading-vcf-data"></a>
</h2>
<p>We start by loading the VCF data using the readVcf function from the
VariantAnnotation package. This data represents structural variant
information. The loaded data is then converted into breakpoint GRanges
notation using the breakpointRanges function.</p>
<p>StructuralVariantAnnotation is built on top of the Bioconductor
package VariantAnnotation. VCF files are loaded using
VariantAnnotation::readVcf, and converted to breakpoint GRanges notation
using breakpointRanges Any non-structural variants such as single
nucleotide variants in a VCF will be silently ignored by
StructuralVariantAnnotation. More information about VCF objects can be
found by consulting the vignettes in the VariantAnnotation package with
browseVignettes(“VariantAnnotation”).</p>
<p>Colo829, is a human cell line primarily derived from skin tissue
affected by melanoma, a type of malignant skin cancer. The cell line is
maintained and distributed by the American Type Culture Collection
(ATCC) under the product name “COLO 829.” Its unique identifier is
Catalog Number CRL-1974.</p>
<p>The Colo829 cell line was established from a 45-year-old male patient
with melanoma, making it an important resource for research and study of
this particular disease. The cells are preserved in a frozen state,
allowing for long-term storage and viability.</p>
<p>These cells are classified as fibroblast-like, which indicates their
morphological and functional characteristics are similar to fibroblasts,
a type of connective tissue cell involved in wound healing and
extracellular matrix production.</p>
<p>With a Biosafety Level of 1, Colo829 is considered safe to handle and
work with under standard laboratory conditions. This makes it accessible
to a wider range of researchers and laboratories.</p>
<p>The primary application of Colo829 is in 3D cell culture experiments,
where it can be used to model melanoma or study interactions between
cancer cells and their surrounding environment in a more physiologically
relevant setting.</p>
<p>Being a human cell line of Homo sapiens origin, Colo829 is a valuable
tool for studying human-specific processes and responses. Its tissue
origin from the skin makes it particularly relevant for melanoma
research, as the disease primarily affects skin cells.</p>
<p>Overall, Colo829 is a well-characterized cell line that holds
significance in the field of melanoma research and offers a versatile
platform for investigating various aspects of cancer biology and
therapeutic development.</p>
<p>See more info about this cell line at <a href="https://www.cellosaurus.org/CVCL_1137" class="external-link uri">https://www.cellosaurus.org/CVCL_1137</a></p>
<div class="sourceCode" id="section-cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">colo829_vcf</span> <span class="op">&lt;-</span> <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">readVcf</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"extdata"</span>,</span>
<span>    <span class="st">"COLO829T.purple.sv.ann.vcf.gz"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"StructuralVariantAnnotation"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">colo829_bpgr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/StructuralVariantAnnotation/man/breakpointRanges.html" class="external-link">breakpointRanges</a></span><span class="op">(</span><span class="va">colo829_vcf</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="section-visualizing-breakpoint-pairs-via-circos-plots">Visualizing breakpoint pairs via circos plots<a class="anchor" aria-label="anchor" href="#section-visualizing-breakpoint-pairs-via-circos-plots"></a>
</h2>
<p>We can visualize breakpoints with a circos plot.</p>
<p>To plot structural variant breakpoints in a circos plot using
shiny.gosling, we need to first prepare the breakpoint GRanges. The
function requires a special column, indicating the end of the link using
GRanges format.</p>
<p>This chunk of code is manipulating genomic ranges represented by the
‘granges’ object named ‘colo829_bpgr’ to create a circular visualization
layout using the ‘biovizBase’ package. The goal is to subset the genomic
ranges relevant to the human genome version ‘hg19sub’ and then convert
these ranges into a circular layout.</p>
<div class="sourceCode" id="section-cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># In this line, a subset of genomic ranges is extracted from the object</span></span>
<span><span class="co"># colo829_bpgr. The subset is obtained by filtering based on the sequence</span></span>
<span><span class="co"># names (seqnames) present in colo829_bpgr that are also found in the</span></span>
<span><span class="co"># sequence levels (seqlevels) of the hg19sub reference genome from the</span></span>
<span><span class="co"># biovizBase package. colo829_bpgr is assumed to be a GenomicRanges object</span></span>
<span><span class="co"># containing genomic regions.</span></span>
<span><span class="va">gr_circos</span> <span class="op">&lt;-</span> <span class="va">colo829_bpgr</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">colo829_bpgr</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="fu">biovizBase</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/biovizBase/man/CRC.html" class="external-link">hg19sub</a></span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># This line updates the sequence levels of the gr_circos object to match</span></span>
<span><span class="co"># those of the hg19sub reference genome. It ensures that the genomic ranges</span></span>
<span><span class="co"># in gr_circos align properly with the reference genome.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">gr_circos</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="fu">biovizBase</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/biovizBase/man/CRC.html" class="external-link">hg19sub</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Here, the metadata columns (mcols) of the gr_circos object are updated</span></span>
<span><span class="co"># with additional information. Specifically, a new column called "to.gr" is</span></span>
<span><span class="co"># added, and its values are assigned using the granges function on the</span></span>
<span><span class="co"># partner genomic ranges of gr_circos.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">gr_circos</span><span class="op">)</span><span class="op">$</span><span class="va">to.gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html" class="external-link">granges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/StructuralVariantAnnotation/man/partner.html" class="external-link">partner</a></span><span class="op">(</span><span class="va">gr_circos</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="section-using-the-granges-object-for-a-circos-plot-using-shiny-gosling">Using the GRanges object for a circos plot using shiny.gosling<a class="anchor" aria-label="anchor" href="#section-using-the-granges-object-for-a-circos-plot-using-shiny-gosling"></a>
</h2>
<div class="section level3">
<h3 id="section-setup-tracks">Setup Tracks<a class="anchor" aria-label="anchor" href="#section-setup-tracks"></a>
</h3>
<div class="section level4">
<h4 id="section-track-1">Track 1<a class="anchor" aria-label="anchor" href="#section-track-1"></a>
</h4>
<p>We will have 3 tracks in this circos plot. The outermost layer
displays the width of a sequence as a bar plot in maroon.</p>
<p>In the track below, we first give it an <code>id</code>. In the
following line, we assign a title to it. This title will be displayed at
the top of the rings. Next, we prepare and pass the data with
<code>track_data_gr</code>, which takes the GRanges object, the
chromosome field names and the genomic field names. We set the mark to
<code>bar</code> as we need a simple bar plot here. We define x and y
via <code>visual_channel_x</code> and <code>visual_channel_y</code>.
Finally, we set the color of the bar plot and the width &amp;
height.</p>
<div class="sourceCode" id="section-cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">track1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_single_track.html">add_single_track</a></span><span class="op">(</span></span>
<span>  id <span class="op">=</span> <span class="st">"track1"</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"Quality"</span>,</span>
<span>  data <span class="op">=</span> <span class="fu"><a href="../reference/track_data_gr.html">track_data_gr</a></span><span class="op">(</span></span>
<span>    <span class="va">gr_circos</span>, chromosomeField <span class="op">=</span> <span class="st">"seqnames"</span>,</span>
<span>    genomicFields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"start"</span>, <span class="st">"end"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  mark <span class="op">=</span> <span class="st">"bar"</span>,</span>
<span>  x <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_x.html">visual_channel_x</a></span><span class="op">(</span>field <span class="op">=</span> <span class="st">"start"</span>, type <span class="op">=</span> <span class="st">"genomic"</span>, axis <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span>,</span>
<span>  y <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_y.html">visual_channel_y</a></span><span class="op">(</span>field <span class="op">=</span> <span class="st">"QUAL"</span>, type <span class="op">=</span> <span class="st">"quantitative"</span>, axis <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span>,</span>
<span>  color <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_color.html">visual_channel_color</a></span><span class="op">(</span></span>
<span>    value <span class="op">=</span> <span class="st">"#BB3114"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  width <span class="op">=</span> <span class="fl">700</span>,</span>
<span>  height <span class="op">=</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="section-track-2">Track 2<a class="anchor" aria-label="anchor" href="#section-track-2"></a>
</h4>
<p>The middle layer plots the start of the sequence to its end.</p>
<p>In this layer, we begin similarly by setting the id and the title of
the layer. Next we pass the same data object as our data will not
change. But other data can be used here. We just need to pass the
appropriate <code>chromosomeField</code> and <code>genomicFields</code>.
We set this as <code>mark = "rect"</code> as we want to fill the ring
with a color if data is a certain value. These are called ideograms. We
set the <code>strokeWidth</code> to 0.5. We set the stroke itself to
have different colors based on the value in the <code>field</code>
argument. Since “REF” is passed as field, and it has 4 values: A, C, G
and T, we set the domain and range accordingly. A corresponds to color
<code>#73A97D</code> and so on. Finally, we set the <code>x</code> and
<code>xe</code>. Since this is a <code>rect</code> ( rectangle ) plot,
it only has <code>x</code> and <code>xe</code> and no
<code>y</code>.</p>
<div class="sourceCode" id="section-cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">track2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_single_track.html">add_single_track</a></span><span class="op">(</span></span>
<span>  id <span class="op">=</span> <span class="st">"track2"</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"REF"</span>,</span>
<span>  data <span class="op">=</span> <span class="fu"><a href="../reference/track_data_gr.html">track_data_gr</a></span><span class="op">(</span></span>
<span>    <span class="va">gr_circos</span>, chromosomeField <span class="op">=</span> <span class="st">"seqnames"</span>,</span>
<span>    genomicFields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"start"</span>, <span class="st">"end"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  mark <span class="op">=</span> <span class="st">"rect"</span>,</span>
<span>  strokeWidth <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_stroke_width.html">visual_channel_stroke_width</a></span><span class="op">(</span></span>
<span>    value <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span>,</span>
<span>  stroke <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_stroke.html">visual_channel_stroke</a></span><span class="op">(</span></span>
<span>    field <span class="op">=</span> <span class="st">"REF"</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"nominal"</span>,</span>
<span>    domain <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"G"</span>, <span class="st">"T"</span><span class="op">)</span>,</span>
<span>    range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"#73A97D"</span>, <span class="st">"#C1BB78"</span>, <span class="st">"#1F5E89"</span>, <span class="st">"#CF784B"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  x <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_x.html">visual_channel_x</a></span><span class="op">(</span>field <span class="op">=</span> <span class="st">"start"</span>, type <span class="op">=</span> <span class="st">"genomic"</span><span class="op">)</span>,</span>
<span>  xe <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_x.html">visual_channel_x</a></span><span class="op">(</span>field <span class="op">=</span> <span class="st">"end"</span>, type <span class="op">=</span> <span class="st">"genomic"</span><span class="op">)</span>,</span>
<span>  width <span class="op">=</span> <span class="fl">700</span>,</span>
<span>  height <span class="op">=</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="section-track-3">Track 3<a class="anchor" aria-label="anchor" href="#section-track-3"></a>
</h4>
<p>Finally, the innermost layer shows the changes in sequences.</p>
<p>For the last layer, we start off again by setting the id and the
title. We set the data with <code>track_data_gr</code> but we now set
the <code>chromosomeField</code> and <code>genomicFields</code> to
<code>to.gr</code> columns. We also set <code>longToWideId</code>. This
will spread the data based on <code>event</code> column. So the new data
will have double the number of columns and will have a new column for
every existing column with the suffix <code>_2</code>. So
<code>to.gr.start</code> will have a corresponding column
<code>to.gr.start_2</code> added to the DB. This is essentially widening
the dataset. We set alignment to overlay, as we want the tracks to be
plotted on top of each other. We then add 2 sub-tracks to this track.
The sub-tracks inherits all the data from the parent track. These tracks
are set with <code>mark = "withinLink"</code> so that the 2 tracks are
linked with each other. And we set <code>x</code>, <code>xe</code>,
<code>x1</code> and <code>x1e</code> to refer to the correct columns in
the data. The first sub-track traces all variations that are not a part
of <code>ch10</code> and the second sub-track traces the ones that are a
part of <code>ch10</code>. The first sub-track will trace on gray color
and the second sub-track will trace in various colors based on its
sequence name. In the end we also specify the width, height and the
stroke width.</p>
<div class="sourceCode" id="section-cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">track3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_single_track.html">add_single_track</a></span><span class="op">(</span></span>
<span>  id <span class="op">=</span> <span class="st">"track3"</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"Highlight similarities of Chr 10 with others"</span>,</span>
<span>  data <span class="op">=</span> <span class="fu"><a href="../reference/track_data_gr.html">track_data_gr</a></span><span class="op">(</span></span>
<span>    <span class="va">gr_circos</span>, chromosomeField <span class="op">=</span> <span class="st">"to.gr.seqnames"</span>,</span>
<span>    genomicFields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"to.gr.start"</span>, <span class="st">"to.gr.end"</span><span class="op">)</span>,</span>
<span>    longToWideId <span class="op">=</span> <span class="st">"event"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  alignment <span class="op">=</span> <span class="st">"overlay"</span>,</span>
<span>  opacity <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_opacity.html">visual_channel_opacity</a></span><span class="op">(</span></span>
<span>    value <span class="op">=</span> <span class="fl">0.4</span></span>
<span>  <span class="op">)</span>,</span>
<span>  tracks <span class="op">=</span> <span class="fu"><a href="../reference/add_multi_tracks.html">add_multi_tracks</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/add_single_track.html">add_single_track</a></span><span class="op">(</span></span>
<span>      dataTransform <span class="op">=</span> <span class="fu"><a href="../reference/track_data_transform.html">track_data_transform</a></span><span class="op">(</span></span>
<span>        type <span class="op">=</span> <span class="st">"filter"</span>, field <span class="op">=</span> <span class="st">"to.gr.seqnames"</span>, oneOf <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"10"</span><span class="op">)</span>, not <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>      <span class="op">)</span>,</span>
<span>      mark <span class="op">=</span> <span class="st">"withinLink"</span>,</span>
<span>      x <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_x.html">visual_channel_x</a></span><span class="op">(</span></span>
<span>        field <span class="op">=</span> <span class="st">"to.gr.start"</span>, type <span class="op">=</span> <span class="st">"genomic"</span></span>
<span>      <span class="op">)</span>,</span>
<span>      xe <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_x.html">visual_channel_x</a></span><span class="op">(</span></span>
<span>        field <span class="op">=</span> <span class="st">"to.gr.start_2"</span>, type <span class="op">=</span> <span class="st">"genomic"</span></span>
<span>      <span class="op">)</span>,</span>
<span>      x1 <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_x.html">visual_channel_x</a></span><span class="op">(</span></span>
<span>        field <span class="op">=</span> <span class="st">"to.gr.end"</span>, type <span class="op">=</span> <span class="st">"genomic"</span></span>
<span>      <span class="op">)</span>,</span>
<span>      x1e <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_x.html">visual_channel_x</a></span><span class="op">(</span></span>
<span>        field <span class="op">=</span> <span class="st">"to.gr.end_2"</span>, type <span class="op">=</span> <span class="st">"genomic"</span></span>
<span>      <span class="op">)</span>,</span>
<span>      stroke <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_stroke.html">visual_channel_stroke</a></span><span class="op">(</span></span>
<span>        value <span class="op">=</span> <span class="st">"lightgray"</span></span>
<span>      <span class="op">)</span>,</span>
<span>      strokeWidth <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_stroke_width.html">visual_channel_stroke_width</a></span><span class="op">(</span></span>
<span>        value <span class="op">=</span> <span class="fl">1</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="fu"><a href="../reference/add_single_track.html">add_single_track</a></span><span class="op">(</span></span>
<span>      dataTransform <span class="op">=</span> <span class="fu"><a href="../reference/track_data_transform.html">track_data_transform</a></span><span class="op">(</span></span>
<span>        type <span class="op">=</span> <span class="st">"filter"</span>, field <span class="op">=</span> <span class="st">"to.gr.seqnames"</span>, oneOf <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"10"</span><span class="op">)</span></span>
<span>      <span class="op">)</span>,</span>
<span>      mark <span class="op">=</span> <span class="st">"withinLink"</span>,</span>
<span>      x <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_x.html">visual_channel_x</a></span><span class="op">(</span></span>
<span>        field <span class="op">=</span> <span class="st">"to.gr.start"</span>, type <span class="op">=</span> <span class="st">"genomic"</span></span>
<span>      <span class="op">)</span>,</span>
<span>      xe <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_x.html">visual_channel_x</a></span><span class="op">(</span></span>
<span>        field <span class="op">=</span> <span class="st">"to.gr.start_2"</span>, type <span class="op">=</span> <span class="st">"genomic"</span></span>
<span>      <span class="op">)</span>,</span>
<span>      x1 <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_x.html">visual_channel_x</a></span><span class="op">(</span></span>
<span>        field <span class="op">=</span> <span class="st">"to.gr.end"</span>, type <span class="op">=</span> <span class="st">"genomic"</span></span>
<span>      <span class="op">)</span>,</span>
<span>      x1e <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_x.html">visual_channel_x</a></span><span class="op">(</span></span>
<span>        field <span class="op">=</span> <span class="st">"to.gr.end_2"</span>, type <span class="op">=</span> <span class="st">"genomic"</span></span>
<span>      <span class="op">)</span>,</span>
<span>      stroke <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_stroke.html">visual_channel_stroke</a></span><span class="op">(</span></span>
<span>        field <span class="op">=</span> <span class="st">"to.gr.seqnames_2"</span>,</span>
<span>        type <span class="op">=</span> <span class="st">"nominal"</span>,</span>
<span>        range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>          <span class="st">"#E79F00"</span>, <span class="st">"#029F73"</span>, <span class="st">"#0072B2"</span>, <span class="st">"#CB7AA7"</span>, <span class="st">"#D45E00"</span>,</span>
<span>          <span class="st">"#57B4E9"</span>, <span class="st">"#EFE441"</span></span>
<span>        <span class="op">)</span></span>
<span>      <span class="op">)</span>,</span>
<span>      strokeWidth <span class="op">=</span> <span class="fu"><a href="../reference/visual_channel_stroke_width.html">visual_channel_stroke_width</a></span><span class="op">(</span></span>
<span>        value <span class="op">=</span> <span class="fl">1.5</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  width <span class="op">=</span> <span class="fl">700</span>,</span>
<span>  height <span class="op">=</span> <span class="fl">200</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="section-final-view">Final View<a class="anchor" aria-label="anchor" href="#section-final-view"></a>
</h4>
<p>Finally we put all the 3 main tracks together along with some
parameters to control the layout style, the alignment and so on.</p>
<div class="sourceCode" id="section-cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_composed_track</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compose_view.html">compose_view</a></span><span class="op">(</span></span>
<span>  title <span class="op">=</span> <span class="st">"Circos"</span>,</span>
<span>  subtitle <span class="op">=</span> <span class="st">"http://circos.ca/intro/genomic_data/"</span>,</span>
<span>  layout <span class="op">=</span> <span class="st">"circular"</span>,</span>
<span>  static <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  spacing <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  centerRadius <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  alignment <span class="op">=</span> <span class="st">"stack"</span>,</span>
<span>  multi <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  tracks <span class="op">=</span> <span class="fu"><a href="../reference/add_multi_tracks.html">add_multi_tracks</a></span><span class="op">(</span><span class="va">track1</span>, <span class="va">track2</span>, <span class="va">track3</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="section-run-app">Run App<a class="anchor" aria-label="anchor" href="#section-run-app"></a>
</h4>
<p>We put all of this in a shiny app.</p>
<div class="sourceCode" id="section-cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ui</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/shiny/man/fluidPage.html" class="external-link">fluidPage</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/use_gosling.html">use_gosling</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/shiny/man/fluidPage.html" class="external-link">fluidRow</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/shiny/man/column.html" class="external-link">column</a></span><span class="op">(</span><span class="fl">6</span>, <span class="fu"><a href="../reference/goslingOutput.html">goslingOutput</a></span><span class="op">(</span><span class="st">"gosling_plot"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/shiny/man/column.html" class="external-link">column</a></span><span class="op">(</span></span>
<span>      <span class="fl">1</span>, <span class="fu"><a href="https://rstudio.github.io/htmltools/reference/builder.html" class="external-link">br</a></span><span class="op">(</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/shiny/man/actionButton.html" class="external-link">actionButton</a></span><span class="op">(</span></span>
<span>        <span class="st">"download_pdf"</span>,</span>
<span>        <span class="st">"PDF"</span>,</span>
<span>        icon <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/shiny/man/icon.html" class="external-link">icon</a></span><span class="op">(</span><span class="st">"cloud-arrow-down"</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">server</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">input</span>, <span class="va">output</span>, <span class="va">session</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">output</span><span class="op">$</span><span class="va">gosling_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renderGosling.html">renderGosling</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/gosling.html">gosling</a></span><span class="op">(</span></span>
<span>      component_id <span class="op">=</span> <span class="st">"component_1"</span>,</span>
<span>      <span class="va">single_composed_track</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/shiny/man/observeEvent.html" class="external-link">observeEvent</a></span><span class="op">(</span><span class="va">input</span><span class="op">$</span><span class="va">download_pdf</span>, <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/export_pdf.html">export_pdf</a></span><span class="op">(</span>component_id <span class="op">=</span> <span class="st">"component_1"</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/shinyApp.html" class="external-link">shinyApp</a></span><span class="op">(</span><span class="va">ui</span>, <span class="va">server</span><span class="op">)</span></span></code></pre></div>
<p><img src="VCF.png" width="500"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="section-session-info">Session Info<a class="anchor" aria-label="anchor" href="#section-session-info"></a>
</h2>
<div class="sourceCode" id="section-cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] vctrs_0.6.3       cli_3.6.1         knitr_1.44        rlang_1.1.1      </span></span>
<span><span class="co">##  [5] xfun_0.40         stringi_1.7.12    purrr_1.0.2       promises_1.2.1   </span></span>
<span><span class="co">##  [9] textshaping_0.3.6 shiny_1.7.5       xtable_1.8-4      jsonlite_1.8.7   </span></span>
<span><span class="co">## [13] glue_1.6.2        rprojroot_2.0.3   htmltools_0.5.6   httpuv_1.6.11    </span></span>
<span><span class="co">## [17] ragg_1.2.5        sass_0.4.7        rmarkdown_2.25    evaluate_0.21    </span></span>
<span><span class="co">## [21] jquerylib_0.1.4   ellipsis_0.3.2    fastmap_1.1.1     yaml_2.3.7       </span></span>
<span><span class="co">## [25] lifecycle_1.0.3   memoise_2.0.1     stringr_1.5.0     compiler_4.3.1   </span></span>
<span><span class="co">## [29] fs_1.6.3          Rcpp_1.0.11       later_1.3.1       systemfonts_1.0.4</span></span>
<span><span class="co">## [33] digest_0.6.33     R6_2.5.1          magrittr_2.0.3    bslib_0.5.1      </span></span>
<span><span class="co">## [37] tools_4.3.1       mime_0.12         pkgdown_2.0.7     cachem_1.0.8     </span></span>
<span><span class="co">## [41] desc_1.4.2</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed with <span class="emoji" data-emoji="heart">❤️</span> by <a href="https://appsilon.com" class="external-link">Appsilon</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
